src/compiler/checker.ts

```diff
@@ -14620,7 +14648,7 @@ namespace ts {
                 else {
                     result = createIntersectionType(typeSet, aliasSymbol, aliasTypeArguments);
                 }
-                intersectionTypes.set(id, result);
+                intersectionTypes.set(id, result as IntersectionType);

```

Comment:
Is this safe? Seems from above that `result` can actually be `Union`.

Reply:
On a first glance, this seems to be so. However, if it is, it was probably incorrect even before this PR, right?

This change is just the result of changing `intersectionTypes` from `Map<string, Type>` to `Map<string, IntersectionType>` - a change that seems logical, however, I inherited it from #39746, so maybe @ShuiRuTian may know the rationale better...

Reply:
I don't think so, given `getIntersectionType` returns `Type`, not `IntersectionType`, and did not need casts beforehand.

Reply:
I've reworked the logic around `intersectionTypes` so that it can just be `Map<string, Type>`.

Reply:
Oops, sorry, it is a mistake. I change the type of `intersectionTypes` to `Map<string, IntersectionType>` just because the type of `unionTypes` is `Map<string, UnionType>`. However, it's proven such infer is too careless and dangerous.